proposals/jvm-expose-boxed.md

```diff
@@ -0,0 +1,268 @@
+
+### No argument constructors
+
+Frameworks like Java Persistence require classes to have a [default no argument constructor](https://www.baeldung.com/jpa-no-argument-constructor-entity-class). We propose to expose a no argument constructor whenever a default value is given to the underlying property of the value class (in addition to the factory methods). Continuing with our example of positive integers, the following code,
+
+```kotlin
+@JvmInline value class PositiveInt(val number: Int = 0) {
+  init { require(number >= 0) }
+}
+```
+
+exposes both the constructor taking a single integer, but also another one without the optional argument.
+
+```kotlin
+class PositiveInt {
+  private synthetic <init>(number: Int, marker: BoxingConstructorMarker)
+
+  public <init>(number: Int): void = <init>(constructor-impl(number), null)
+  public <init>(): void = <init>(0)
+}
+```
+
+### JVM value classes
+
+It is not a goal of this KEEP to decide what should happen with Kotlin value classes once the JVM exposes that feature ([JEP-401](https://openjdk.org/jeps/401)). The compilation scheme presented above follows the guidelines for _Migration of existing classes_, so it should be possible to neatly migrate to a JVM value class without further changes.
+
+### Other design choices
+
+**Put it under a flag or annotation**, in other words, move to the new generation scheme only when some annotation `@JvmExposeFactory` or compiler flag is present. In this case, we could not find a realistic scenario where this expose would be counter-productive; in the worst case in which you expose the new constructor but not operations you just have a way to create useless values.
+
+**Exposing a factory method**: instead of exposing the constructor, use `PositiveInt.of(3)`. This option is nowadays quite idiomatic in the Java world (for example, in `List.of(1, 2)`), but not as much as constructors. Using a factory method also has the drawback of possible clashes for the name of the factory method, which we avoid when using the constructor.
+
+## Expose operations and members
+
+The current compilation scheme transforms every operation where a value class is involved into a static function that takes the unboxed variants, and whose name is mangled to prevent clashes. This means those operations are not available for Java consumers. We propose introducing a new `@JvmExposeBoxed` annotation that exposes a variant of the function taking and returning the boxed versions instead (if more than one argument or return type is a value class, the aforementioned variant uses the boxed versions for _all_ of them). We call it the _boxed variant_ of the function.
+
+```kotlin
+annotation class JvmExposedBoxed(val jvmName: String = "")
+```
+
+The `@JvmExposeBoxed` annotation may be applied to a declaration, or a declaration container (classes, files). In the latter case, it should be taken as applied to every single declaration within it.
```

Comment:
Is it allowed on open members? If yes, it kind of makes [KT-31420](https://youtrack.jetbrains.com/issue/KT-31420) partially supported which is a bit suspicious. In particular, we should probably examine all problems keeping us from implementing KT-31420 and see how they affect `@JvmExposeBoxed`.